domain/src/com/thoughtworks/go/domain/materials/svn/SvnMaterialUpdater.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.domain.materials.svn;
+
+import com.thoughtworks.go.config.materials.svn.SvnMaterial;
+import com.thoughtworks.go.domain.BuildCommand;
+import com.thoughtworks.go.domain.materials.Revision;
+import com.thoughtworks.go.domain.materials.RevisionContext;
+import com.thoughtworks.go.util.command.UrlArgument;
+import org.apache.commons.lang.StringUtils;
+
+import java.io.File;
+import java.util.ArrayList;
+
+import static com.thoughtworks.go.domain.BuildCommand.*;
+import static java.lang.String.format;
+
+public class SvnMaterialUpdater {
+   private SvnMaterial material;
+
+   public SvnMaterialUpdater(SvnMaterial material) {
+      this.material = material;
+   }
+
+   public BuildCommand updateTo(String baseDir, RevisionContext revisionContext) {
+      Revision revision = revisionContext.getLatestRevision();
+      String workingDir = material.workingdir(new File(baseDir)).getPath();
+      UrlArgument url = material.getUrlArgument();
+
+      return compose(
+              echoWithPrefix(format("Start updating %s at revision %s from %s", material.updatingTarget(), revision.getRevision(), url.forDisplay())),
+              secret(url.forCommandline(), url.forDisplay()),
+              secret(material.getPassword(), "*********************"),
+              secret(material.currentPassword(), "*********************"),
+              cleanupAndUpdate(workingDir, revision).setTest(shouldDoCleanupAndUpdate(workingDir)),
+              freshCheckout(workingDir, revision).setTest(isNotRepository(workingDir)),
+              freshCheckout(workingDir, revision).setTest(test("-nd", workingDir)),
+              freshCheckout(workingDir, revision).setTest(repoUrlChanged(workingDir)),
```

Comment:
It felt, to me, that this could checkout 3 times in certain cases. But, after some thought, it looks like it won't, because of the order of those three statements. I think it would be nice to change this to:

```java
freshCheckout(workingDir, revision).setTest(not(shouldDoCleanupAndUpdate(workingDir))),
```

But, that has a problem too, where the second check is affected by the previous action. So, your approach is fine.